### PR TITLE
Optional relabelling of nodes & edges with `Annotation.UNKNOWN` label

### LINKED ISSUE #185: Include optional argument to relabel `Annotation.UNKNOWN` label
In cases of an **exhaustive** annotation, _i.e._ when all real objects are annotated, one may want to set all `Annotation.UNKNOWN` labels to negative. This also enhances training & helps with to speed the process up.

### LINKED ISSUE #197: Handle montage display with only 2 annotation classes

### LINKED ISSUE #201: Print out image / graph node & edge statistics if desired
Include an optional argument in `ImageGraphDataset` to print out the image statistics:

https://github.com/alan-turing-institute/grace/blob/9f744956136bd039e4544d2cc64329917ac8638f/grace/io/image_dataset.py#L20-L22

https://github.com/alan-turing-institute/grace/blob/9f744956136bd039e4544d2cc64329917ac8638f/grace/io/image_dataset.py#L40-L41

### LINKED ISSUE #204: Use `logging` rather than `print` for user feedback

### PR DESCRIPTION
**Pull request by KristinaUlicna**
If the user decides to do so, relabel all GT annotations from `Annotation.UNKNOWN` to `Annotation.TRUE_NEGATIVE`, to increase the number of data points to train the GCN on. Particularly good for **exhaustive** annotation.

- [x] Resolves #185 
- [x] Resolves #201 
- [x] Resolves #197 
- [x] Resolves #204

### CODE REVIEW

File: grace/io/image_dataset.py [hunk 1]
```diff
@@ -86,6 +123,41 @@ def __getitem__(self, idx: int) -> tuple[torch.Tensor, dict]:
         return image, target
 
 
+def relabel_unknown_labels(G: nx.Graph):
```
Comment (crangelsmith) [inline grace/io/image_dataset.py:126]: This is perfectly ok, but I just wanted to highlight that there is a networkx method to do this maybe more efficiently? https://networkx.org/documentation/stable/reference/generated/networkx.relabel.relabel_nodes.html

File: grace/training/train.py [hunk 2]
```diff
@@ -12,6 +13,12 @@
 
 from torch.utils.tensorboard import SummaryWriter
 
+logging.basicConfig(
```
Comment (crangelsmith) [inline grace/training/train.py:16]: I think you only have to do this config once, in the run and it is saved for everywhere else the logger is called. In affinityvae we do it in the [run.py script](https://github.com/alan-turing-institute/affinity-vae/blob/develop/run.py)

File: grace/io/image_dataset.py [hunk 3]
```diff
@@ -74,9 +95,25 @@ def __getitem__(self, idx: int) -> tuple[torch.Tensor, dict]:
             self.image_reader_fn(img_path), dtype=torch.float32
         )
         grace_dataset = read_graph(grace_path)
+        graph = grace_dataset.graph
+
+        # Print original graph label statistics:
+        if self.verbose is True:
```
Comment (crangelsmith) [inline grace/io/image_dataset.py:101]: Why using if <value> is True, instead of if <value>? Same question for the lines below.

**Comment by crangelsmith** [comment]
I'm trying to run the code, but I get a fail with the following error:
``` 
File "/Users/crangelsmith/Projects/grace/grace/models/classifier.py", line 72, in forward
    embeddings, p=self.dropout, training=self.training
    ^^^^^^^^^^
UnboundLocalError: cannot access local variable 'embeddings' where it is not associated with a value
```
I think it is due to the embeddings variable in this line not being defined is the condition in line 66 is not met. 
https://github.com/alan-turing-institute/grace/blob/a4d98ef9320123ef81ba57d4feace57fe8fdac68/grace/models/classifier.py#L69

File: grace/io/image_dataset.py (hunk 1 above)
Comment (KristinaUlicna) [reply grace/io/image_dataset.py:126]: I've looked at this function, but the mapping allows you to use new names for the node_idx, not for node_attributes. Here, what I need to do is to iterate over all nodes & edges to relabel their `GraphAttrs.GROUND_TRUTH` to `Annotation.TRUE_NEGATIVE`. Is there a handy function to do this simpler?

File: grace/training/train.py (hunk 2 above)
Comment (KristinaUlicna) [reply grace/training/train.py]: Addressed now, logger config specifies in the `run.py` exclusively 😃

**Comment by KristinaUlicna** [comment]
> I'm trying to run the code, but I get a fail with the following error:
> 
> ```
> File "/Users/crangelsmith/Projects/grace/grace/models/classifier.py", line 72, in forward
>     embeddings, p=self.dropout, training=self.training
>     ^^^^^^^^^^
> UnboundLocalError: cannot access local variable 'embeddings' where it is not associated with a value
> ```
> 
> I think it is due to the embeddings variable in this line not being defined is the condition in line 66 is not met.
> 
> https://github.com/alan-turing-institute/grace/blob/a4d98ef9320123ef81ba57d4feace57fe8fdac68/grace/models/classifier.py#L69

Temporarily fixed now, proper fix for option to train a linear-only classifier upcoming in #183 🚀

File: grace/io/image_dataset.py (hunk 1 above)
Comment (crangelsmith) [reply grace/io/image_dataset.py:119]: I see, i think then your approach is the best one :)

File: grace/io/image_dataset.py (hunk 3 above)
Comment (crangelsmith) [reply grace/io/image_dataset.py:94]: No preference, I only thought that in the installation of the class, the type is already defined, this step might not be necessary but maybe there was another reason.

File: grace/io/image_dataset.py [hunk 4]
```diff
@@ -30,6 +34,12 @@ class ImageGraphDataset(Dataset):
         Transformation added to the images and targets
     image_filetype : str
         File extension of the image files
+    keep_unknown_labels : bool
```
Comment (crangelsmith) [inline grace/io/image_dataset.py:37]: is this configured somewhere or only here? shouldn't be part of the config file so the user can decide on the fly how to process it?

**Comment by crangelsmith** [comment]
> > I'm trying to run the code, but I get a fail with the following error:
> > ```
> > File "/Users/crangelsmith/Projects/grace/grace/models/classifier.py", line 72, in forward
> >     embeddings, p=self.dropout, training=self.training
> >     ^^^^^^^^^^
> > UnboundLocalError: cannot access local variable 'embeddings' where it is not associated with a value
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think it is due to the embeddings variable in this line not being defined is the condition in line 66 is not met.
> > https://github.com/alan-turing-institute/grace/blob/a4d98ef9320123ef81ba57d4feace57fe8fdac68/grace/models/classifier.py#L69
> 
> Temporarily fixed now, proper fix for option to train a linear-only classifier upcoming in #183 🚀

Cool, it runs now without issues.

**Review by crangelsmith — APPROVED** [review]
I've run the code with the relabelling annotation on and off and seems to run well. My only comment left is on making the relabeling a configurable option for the user, but besides that I think is ready to merge.

File: grace/io/image_dataset.py (hunk 4 above)
Comment (KristinaUlicna) [reply grace/io/image_dataset.py:37]: Very good point, adding this into #207 ! For now, it can be specified in the `run.py` directly but nowhere else.

**Pull request merged**